### ISSUE TITLE
Add API for requesting permission to receive device motion / orientation events

### ISSUE THREAD

**Issue by cdumez** [w3c/deviceorientation#57]
Due to the privacy issues related to this API (e.g. [https://www.wired.com/story/mobile-websites-can-tap-into-your-phones-sensors-without-asking/](https://www.wired.com/story/mobile-websites-can-tap-into-your-phones-sensors-without-asking/) and [https://arxiv.org/pdf/1605.08763.pdf](https://arxiv.org/pdf/1605.08763.pdf)), we are considering adding to WebKit a permission dialog in order to ask the user if they want to expose their device orientation/motion to the Website.

However, because trackers already register such event listeners on top sites, we are worried about the risk of over-prompting. We wouldn't want this prompt to start showing on a lot of top-sites.

To address the issue, we'd like to propose adding a new API allowing the page's script explicitly ask for permission to access the device / motion. We think we should also require a user-gesture to call this new API.

It could look something like this (similar to Notification API):
```
[Exposed=(Window)]
interface DeviceOrientation {
  [Exposed=Window] static Promise<boolean> requestPermission();
};
```

What are your thoughts?

**Comment by reillyeon** [comment]
This specification is superseded by the [Generic Sensors](https://w3c.github.io/sensors/) family of specifications, which include integration with the [Permissions API](https://w3c.github.io/permissions/). I think that this specification should be updated with a similar integration by requiring `"accelerometer"` and `"gyroscope"` permissions for `deviceorientation` and `devicemotion` events and additionally `"magnetometer"` permission for `deviceorientationabsolute` events. This would match the existing [Orientation Sensor](https://w3c.github.io/orientation-sensor/), [Accelerometer](https://w3c.github.io/accelerometer/), [Gyroscope](https://w3c.github.io/gyroscope/) and [Magnetometer](https://w3c.github.io/magnetometer/) specifications which provide access to the same or similar data.

Is there any plan to implement `navigator.permissions.query()` in WebKit?

Blink is currently considering similar changes to start notifying users or prompting when sites request access to device sensors. The plumbing is already implemented to manage these permissions however they are still granted by default.

**Comment by cdumez** [comment]
@reillyeon I may be missing something but where is that API that JS actually calls to request permission for something? `navigator.permissions.query()` seems to only query, not request.

There is a section in the Permissions API spec about "requesting more permission" but it seems to merely be an algorithm, no actual API for it.

What we want in WebKit is really an API to ask for permission to access device orientation. Without such permission request, the content would not be getting the events at all.

**Comment by reillyeon** [comment]
That is correct, I forgot that the `navigator.permissions.request()` method has been removed in favor of each specification integrating its own method for requesting more permission, but calling into the algorithm you mention. A cursory look at `query()` made me think that that had been integrated into it instead.

Given that the Generic Sensors specifications I linked to above should be updated to include a `requestPermission()` method similar to [Notification.requestPermission()](https://notifications.spec.whatwg.org/#dom-notification-requestpermission).

As the DeviceOrientation API does not provide an object in the global scope like AbsoluteOrientationSensor or Notification (except perhaps DeviceOrientationEvent) what do you think about not adding a method specific to this API but instead relying on the ones that need to be added to the newer specifications?

My recollection is that WebKit does not implement the `deviceorientationabsolute` event and so implementing a single `RelativeOrientationSensor.requestPermission()` method would suffice for your purposes.

**Comment by cdumez** [comment]
I think relying on a Permissions API from another spec would probably work (although, I'd have to see what it looks like first). That said, it would be bit confusing to use Permissions API from another (newer) API and predicate using the (older) API on it.

How about for the 'devicemotion' events, would it be a `LinearAccelerationSensor.requestPermission()` ?

Honestly, the fact that the names do not match would be super-confusing IMO.

I get that the DeviceOrientation API does not have a global scope object at the moment but it seems like it could have one. Or I actually like the idea of adding static requestPermission() methods to the existing DeviceOrientationEvent / DeviceMotionEvent interfaces.

**Comment by anssiko** [comment]
`navigator.permissions.request()` is specified in https://wicg.github.io/permissions-request/

@jyasskin probably knows the status of that work.

**Comment by cdumez** [comment]
So if I understand correctly, it would look like:
navigation.permissions.request({ name: “gyroscope” }); // For DeviceOrientationEvents
navigation.permissions.request({ name: “accelerometer” }); // For DeviceMotionEvents

This does not look bad actually.

**Comment by reillyeon** [comment]
You need both `"gyroscope"` and `"accelerometer"` for both so it would look like:

```js
Promise.all([navigation.permissions.request({ name: “accelerometer” }),
             navigation.permissions.request({ name: “gyroscope” })])
    .then(results => {
      ...
    });
```

**Comment by jyasskin** [comment]
I think https://github.com/w3c/permissions/issues/158 is the most recent discussion about `permissions.request(...)`, and it links to the older discussions. Firefox folks have generally been opposed, Chrome's been in favor, and Edge and Safari have ignored it. I haven't been paying a lot of attention in the last year so can't really help guide here.

**Comment by cdumez** [comment]
@annevk Based on https://github.com/w3c/permissions/issues/158, it looks like Mozilla considered gating the old deviceorientation / devicemotion events behind a permission as well. Did you end up doing it or do you still have plans to do so?

**Comment by annevk** [comment]
@cdumez we haven't made progress on this yet I believe, but I strongly suspect we're still interested. Perhaps not in a generic `request()` method though, but something specific for this API. Either that, or seeing if we can remove the API altogether (we don't expose it so widely I think).

cc @ehsan

**Comment by cdumez** [comment]
Yes, I discussed it internally a bit and I do not think we are interested in the generic permissions.request().

A few things we could do:
- add a permission request api to this spec
- drop the api altogether but there are some legit use cases so it would be unfortunate.
- maybe require a user gesture when adding the event listener (and then we’d show a prompt). Avoids over prompting and adding any api.

**Comment by richtr** [comment]
From the study referenced in that Wired article (PDF paper: https://sensor-js.xyz/webs-sixth-sense-ccs18.pdf) the recommendations given to browser vendors are as follows:

> * W3C’s recommendation for disabling sensor access on crossorigin
> iframes [80] will limit the access from untrusted thirdparty
> scripts and is a step in the right direction. However,
> Safari and Firefox are the only two browsers that follow this
> recommendation. Our measurements indicate that scripts
> that access sensor APIs are frequently embedded in crossorigin
> iframes (67.4% of the 31 444 cases). This shows that
> W3C’s mitigation would be effective at curbing the exposure
> to untrusted scripts. Allowing sensor access on insecure
> pages is another issue where browsers do not follow the
> W3C spec: all nine browsers we studied allowed access to
> sensors on insecure (HTTP) pages.
> 
> * Feature Policy API [16], if deployed, will allow publishers
> to selectively disable JavaScript APIs. Publisher may disable
> sensor APIs using this API to prevent potential misuses by
> the third-party scripts they embed.
> 
> * Provide low resolution sensor data by default, and require
> user permission for higher resolution sensor data.
> 
> * To improve user awareness and curb surreptitious sensor
> access, provide users with a visual indication that the sensor
> data is being accessed.
> 
> * Require user permission to access sensor data in private
> browsing mode, limit resolution, or disable sensor access all
> together.
> 
> [80] Rich Tibbett, Tim Volodine, Steve Block, and Andrei Popescu. 
> 2018. DeviceOrientation event specification. 
> https://w3c.github.io/deviceorientation/
> 
> [16] Ian Clelland. 2017. Feature policy: Draft community group 
> report. 
> https://wicg.github.io/feature-policy/.

This specification could instead be updated to include normative text that limits the sampling rate for `deviceorientation`, `deviceorientationabsolute` and `devicemotion` events to less than or equal to 60 Hz as effective mitigation to fingerprinting and other security and privacy issues.

At that point, a permissions API could then be introduced for accessing higher frequency sensor data from devices. Or, just require higher frequency sensor data to be obtained only via the Generic Sensors API going forward.

Incremental improvements based on these research recommendations seem possible. Putting all access to these events behind a permission dialog seems like a Really Big Gun to start with. It may still be required to eventually look at that approach but for an API that has been available for 8+ years this could be extremely damaging for legitimate use cases given its widespread adoption.

**Comment by cdumez** [comment]
This would not be acceptable on WebKit-side. Limiting sampling rate for those events would mitigate fingerprinting but not some of the other privacy issues (e.g. The site can tell if the user is standing or laying in bed).

We'd require a user gesture / permission for any of those events. We could also disable API altogether as @annevk mentioned.

**Comment by richtr** [comment]
Determining which way the user is pointing their device is the _entire purpose_ of this API. 

Enabling usage for legitimate sites while preventing fingerprinting seems to be the objective of this proposal based on the contents and research in the cited studies.

The studies you referenced to justify this addition do not recommend putting this behind permission prompts as a first course of action. Those studies present many other options that could be considered.

Why the immediate and full rejection of all recommendations from _your own references_?

**Comment by cdumez** [comment]
I get that this is the purpose of the API but doing it without the user’s consent is another matter.

**Comment by reillyeon** [comment]
I am in favor of adding `requestPermission()` methods to `DeviceOrientationEvent` and `DeviceMotionEvent` (as well as the equivalent Generic Sensors APIs). Acknowledging WebKit's concerns I think it is reasonable for an implementation to choose to completely deny permission for these APIs by default if they wish to. Blink's current plans are to,

1. Implement a permissions API allowing a site to request a particular level of precision and/or frequency.
2. Allow access to the existing level of precision and frequency by default.
3. Notify the user that a site is using the device sensors and allow them to revoke permission.

(2) is necessary to maintain compatibility with existing sites and the existing information leaks are partially mitigated by (3). (1) provides a framework for enabling (a) future use cases that require more data and (b) a move towards reducing the amount of data available without a permission in a gradual fashion in order to allow sites to adapt.

**Comment by richtr** [comment]
In summary, changes could be made immediately to implementations by obfuscating, quantizing and/or reducing the rate of sensor event data emitted without impacting the existing API footprint or validity/conformity of implementations to this specification. We could also immediately add security and privacy considerations to this effect to this specification based on these emergent research studies since last publication. We can then continue to discuss additional requirements / API changes as required, hopefully with a view to reducing the impact of those changes on current legitimate consumers of this API.

**Comment by annevk** [comment]
@richtr you have two-out-of-three implementers saying that's not enough. That doesn't seem like a good summary to me.

**Comment by richtr** [comment]
There are many softer approaches to be explored here as opposed to implementing a black-and-white “on/off” switch straight out of the gate.

Following your current course will effectively destroy nascent VR/AR/360/gaming companies and industries and pretending otherwise is silly.

I want to avoid a cliff-edge scenario where access to these sensors would change overnight without proper planning — particularly in browsers where no platform-alternative browser engines exist. 

For that purpose there are many many intermediate steps, recommended in the cited papers - before you need to implement on/off switches for an 8+ year old API. Jumping straight for that approach is incredibly hostile to web developers. Have you evaluated the full impact that would have?

That is the purpose of the summary above.

**Comment by cdumez** [comment]
I believe @geoffreygaren’s comment above explains why we want a permission api and user consent in WebKit. For us, it is either this or disabling the API by default until we can make it safe. The other proposals do not address all our concerns as explained by Geoff.

**Comment by richtr** [comment]
> For us, it is either this or disabling the API by default until we can make it safe.

All content relying on the API will break overnight. No replacement sensor access is provided on WebKit (Generic Sensors API). No alternative browser engines are allowed on iOS so there is no way to keep any applications relying on those APIs running. No deprecation strategy is planned warning developers of disruption or revocation of sensor data before they land in Safari. A cliff-edge fall-off of usage is the correct way to modify an extensively used 8+ year old web platform API.

Browser-based content and embedded web view content in native applications will cease operation in all of the following use cases:

* Games and gaming
* Navigation
* Virtual Reality
* Augmented Reality
* 360 viewers (videos, panoramas)
* 3D scanning
* Gesture recognition
* Step counters
* Physical activity monitoring
* Sleep monitors 

Multiple proposals exist to make the emitted motion sensor data both a.) still useful for legitimate use cases and b.) resistant to raised security and privacy concerns by refining/limiting the data emitted in different ways. Those solutions will not considered further because the effectiveness of those solutions needs further investigation/research even though they are cited as potentially resolving all current security and privacy concerns by the original investigators. Instead we will modify the API invocation to require permissions/user-gesture or just remove the whole API because it is easier, less implementation work and resolves all security and privacy concerns because then there are no use cases or usage anymore.

Now contrast this to [the alternative approach/summary I posted above](https://github.com/w3c/deviceorientation/issues/57#issuecomment-448533631).

**Comment by zchrykng** [comment]
> Browser-based content and embedded web view content in native applications will cease operation in all of the following use cases:
>
>* Games and gaming
>* Navigation
>* Virtual Reality
>* Augmented Reality
>* 360 viewers (videos, panoramas)
>* 3D scanning
>* Gesture recognition
>* Step counters
>* Physical activity monitoring
>* Sleep monitors

How many of these things are implemented as web apps/views? And why would requiring them to ask permission be that big of a deal? There is exactly zero people who should be able to access any of this data without asking for permission, in my opinion. And I honestly couldn't care less how many business models or apps would be affected by having to ask permission. If the app is scared of asking for permission first, I don't want them having my data anyway.

**Comment by eligrey** [comment]
Companies like UnifyID use gyroscope data for user gait profiling. I would prefer if this data required permission to access.

**Comment by Codeusa** [comment]
Sacrificing UX in the name of a slight privacy increase is regressive at best. Not only will most end-users not understand what granting the permission actually entails, it just creates friction when trying to build solutions that “just work.”

In HTML5 eBook reader I created  I actually leverage device orientation to determine when someone is laying down to suggest turning on night mode at certain hours.

**Comment by anssiko** [comment]
According to an [article](https://medium.com/@firt/pwas-on-ios-12-2-beta-the-good-the-bad-and-the-not-sure-yet-if-good-a37b6fa6afbf#1e1f) discussing iOS 12.2 beta features:
>* Accessing sensor APIs is now disabled by default, so no accelerometer, gyroscope or magnetometer APIs are working unless the user enables a Settings option under “Safari”

Assuming this beta feature will ship in stable, it will diverge from [Blink's plan](https://github.com/w3c/deviceorientation/issues/57#issuecomment-447430709) to allow access to the existing level of precision and frequency by default.

A possible impact on iOS users is that websites start to show bespoke modal dialogs to iOS Safari users that ask to enable the "Motion & Orientation Access" option in iOS Settings (similarly to those "Add to homescreen" bespoke dialogs).

That seems like a suboptimal UX for iOS users, so I'm interested in finding a consensus position among implementers that would enable a better UX for all implementers. If `requestPermission()` is a fix all implementers can live with, we can make a case that it is needed to "fix a remaining known issue" in the API and see if anyone has concerns, and spec it. As said, the charter has some leeway in this regard, so we could likely pull that off with all implementers on board.

Seeking feedback from @cdumez and @reillyeon (with his Blink hat on).

**Comment by larsgk** [comment]
If the aim is to cripple/disable 3rd party ads/scripts to access sensor data for evil intentions, using Feature Policy seems to be a good way forward.  Has this been considered?  If we want to prevent the main page from accessing sensors, we have to consider how requiring installation of a native app would provide better assurance - if any at all.

In addition, permissions could be granted when installing the web application (to home screen/pwa).  Has this been considered?

If too many features require user approval or flipping switches in the global settings, the effect will most likely disappear (e.g. like most people run as Administrator on Windows as the alternative is just too cumbersome).

**Comment by cdumez** [comment]
@anssiko What sort of feedback are you looking for? I filed this issue so requestPermission() is definitely sometime we want and it would definitely be a better situation than having the API disabled by default. @annevk was also on board with such API, so that's 2 implementers.
If Blink wants to keep providing access to the API without user consent, this is fine, the requestPermission() could always be a no-op for them.

@larsgk FYI, third-party content already does not have access to the API in WebKit so this discussion is about first party for us. Also note that the whole reason we want a permission API is to avoid over-prompting the user since there is a lot of content out there already using the API but not actually requiring it (e.g. tracking). With regards to Web Applications (PWA), we'll need to discuss this internally, this is more of a system policy and I do not think this should be spec'd.

**Comment by anssiko** [comment]
@cdumez thanks for confirming your position. That’s what I meant.

@reillyeon would you be fine with the proposed `requestPermission()` that could in your case be no-op or return a promise that immediately resolves?

All - any concerns with the proposal, please let us know.

**Comment by griffin2000** [comment]
So what is the timeline on adoption of both the new Sensor API and the new permission features?  Turning off this feature on IOS completely breaks our product! This is part of the web standard that we were using, and we've put a great deal of effort into developing for.  To have it suddenly stop working is incredibly frustrating.

**Comment by reillyeon** [comment]
@anssiko, I am positive on this. Blink's implementation will start as a no-op that immediately resolves with `"granted"` but I agree that it is important for it to be present nonetheless.

@cdumez, can you put together a draft of this change in a PR so we can discuss the specifics on how this will be specified? From the preceding discussion it sounds like we have settled on extending the `DeviceOrientationEvent` and `DeviceMotionEvent` interfaces with something like,

```
enum PermissionState {
  "granted",
  "denied",
};

[Exposed=(Window)]
interface DeviceOrientationEvent {
  static Promise<PermissionState> requestPermission();
};

[Exposed=(Window)]
interface DeviceMotionEvent {
  static Promise<PermissionState> requestPermission();
};
```

These are a specialization of the so-far unspecified `requestPermission()` method that should be added to the Generic Sensor API that will likely have parameters to control things like requested sampling frequency.

**Comment by cdumez** [comment]
@reillyeon Yes, your proposal looks good to me. With regards to the PR, I'll try and make one soon but this is a busy period for me.

**Comment by rigel-prime** [comment]
Rigel from 8th Wall here - I updated an iPhone 8 Plus to iOS 12.2b today and noticed it broke several sites due to the issues laid out above. I filed a radar earlier and posted an openradar here: https://openradar.appspot.com/47645367 

Can we expect by the time iOS 12.2 is released publicly, to have had sufficient lead time to update our software to accommodate any of the proposed changes so that none of our customers' websites break? There are several high profile VR/AR projects currently in production that this would directly affect. Thank you.

**Comment by anssiko** [comment]
(Related: I submitted a PR #62 that converts the spec source from its custom HTML format into Bikeshed format to make it easier to edit the spec as well as to publish updates using the W3C's publication system Echidna. @cdumez when you start working on the PR, I recommend you base it off of the Bikeshed source.)

**Comment by TimvanScherpenzeel** [comment]
Even though I agree with the general premise I have a few concerns:

The existing VR / AR / 360 camera viewers / experiences will break (similar to the autoplay audio policy did on Chrome). If certain permissions have existed on the web for a long time you cannot remove them without breaking anything.

I'm not sure if I agree with the proposal of @griffin2000 regarding a pop-up asking for permissions. I dread going to any news websites because as soon as the page loads I get bombarded with permission requests and I refuse them 99.9% of the time.

If we want the web to be able to compete with native applications (think seamless AR / VR) developers need powerful tools and be able to create a user experience that does not feel like installing an app every-time to go to a website. In my opinion the web should feel like an open space to surf and explore, not a walled off section with border security.

How exactly are advertisers exploiting these sensors in a way that is deemed so disastrous and unsafe that the rest of the people who try to push the web into new creative directions have to suffer?

**Comment by richtr** [comment]
I'm exasperated by the level of cognitive dissonance in this discussion.

The 3 recommendations to resolve all fingerprinting or key logging ability in this API and to prevent a breakdown of all existing web developer code are as follows:

* **Obfuscation**: Add a random constant in the range [-0.05, 0.05] to all `deviceorientation` data and [-0.005, 0.005] to all `devicemotion` data. Optionally adjust that value during idle periods of inactivity for dynamic obfuscation (when e.g. the device is placed on a table).

* **Quantization**: Round all sensor data to a discrete set of values. For `deviceorientation` round the data emitted to e.g. 2 decimal places. For `devicemotion` round the data emitted to e.g. 4 decimal places.

* **Frequency**: Optionally, reduce sensor data frequency from 60Hz to e.g. 30Hz.

Why are these solutions, that do not impact the footprint of this API, still not being seriously considered? Even if that turns out to be an interim solution it presents a more positive development of this API than the disruptive, backward-incompatible changes being considered.

How would these updates not strike the correct pragmatic balance of resolving all privacy concerns while protecting access for legitimate users?

If these low-impact fixes for the original reported issue continue to be ignored this is what a responsible deprecation process for established Web APIs looks like: https://developers.google.com/web/updates/2018/09/chrome-70-deps-rems#deprecation-policy

Nobody has ruled out a permissions approach for this API if that is the least-worst acceptable solution. I'm just surprised that there is such little interest in progressively experimenting and fixing issues according to the recommendations of all studies in to the issues.

**Comment by reillyeon** [comment]
@richtr The problem with those recommendations is that they still break existing applications but in more subtle ways. As outlined in [my previous comment](https://github.com/w3c/deviceorientation/issues/57#issuecomment-447430709) before we can realistically restrict the fidelity of the data available via this or the Generic Sensors API there needs to be a mechanism for developers to express to the user agent the level of data fidelity required for their particular application. The proposed `requestPermission()` amendment to this API, and the additional parameters to the Sensor constructor in the Generic Sensors API, are that mechanism.

**Comment by othermaciej** [comment]
We'd adopt the `requestPermission()` proposal here in Safari if there is general agreement on it.

Regarding @richtr's [comment](https://github.com/w3c/deviceorientation/issues/57#issuecomment-459302243), we consider this to be a privacy and security hole that is being actively exploited in the wild. Given that, we don't think it is acceptable to do the slow deprecation path or to experiment with different forms of partially limiting the data to see if it works. If the stakes were lower, we would not feel the need to aggressively remove.

**Comment by erikmchut** [comment]
@richtr - I'd be concerned that reducing the quality of the sensor data would significantly degrade the quality of existing AR and VR uses of this standard API. In these webapps, (many of which already request camera permissions), an additional devicemotion permission would be much preferable than materially affecting tracking quality or breaking synchronicity with the camera.

**Comment by richtr** [comment]
@erikmchut Your points are valid wrt VR/AR usage. These changes would increase "motion sickness" considerably. 

However, the DeviceOrientation Events API is not the correct API to use to implement VR/AR. The fact it is used on WebKit-based browsers is because it currently provides the only way to access sensor data in those browsers. 

There are other/better APIs for Web VR/AR implementations available for implementation:

https://immersive-web.github.io/webvr/spec/1.1/
https://www.w3.org/TR/gyroscope/

The issue is that these APIs are not currently supported in WebKit. Instead of implementing those APIs, the proposals here suggest we instead modify/remove this existing API.

What I proposed above that we make `deviceorientation` and `devicemotion` un-fingerprintable and un-key-loggable by making it a source for lower resolution data right now. A lot of non-VR/AR use cases are also at risk here and, on implementation of the above specifications we would have the following scenario:

1. Low-resolution `deviceorientation` and `devicemotion` data, that eliminates fingerprinting or key-logging abilities remains available by default without permissions.
2. High-resolution sensor data is available via both the WebVR API and the Generic Sensors API (see above for links) with user permissions being acquired first.

This represents a way to keep things moving on a web page that need to move (e.g. motion parallax effects, custom gesture recognition, CSS animations attached to device pose, etc) with a path for developers to "upgrade" that default low-level data access to something suitable with the permission of users.

> If the stakes were lower, we would not feel the need to aggressively remove.

This is very damaging to every stakeholder concerned and a shock to the web standards system. I'm not aware of any other such widely used APIs receiving the same treatment previously. As such it feels unprecedented in the history of Web Standards to take such a drastic course of action instead of progressively fixing the issue.

**Comment by geoffreygaren** [comment]
@richtr You keep asserting that obfuscation, quantization, and frequency reduction solve the privacy and consent problem.

But the only sources cited in this conversation acknowledge that all such efforts so far have failed to solve the privacy and consent problem.

Regarding fingerprinting, Das-Borisov-Acar-Pradeep acknowledge that users who want privacy still need to “disable sensor access all together” (Section 7, DISCUSSION AND RECOMMENDATIONS”), and they “recommend browser vendors to rethink the risks of exposing sensitive sensors without any form of access control mechanism in place” (Section 8, LIMITATIONS). Das-Borisov-Chou-Mughees acknowledge that their best effort to curtail fingerprinting still has a Random Forest F-score of 26% (Section 6.2, Effectiveness of Countermeasures).

Regarding key logging and other forms of eavesdropping, nobody has suggested that obfuscation, quantization, or frequency reduction is a mitigation at all. In fact, Das-Borisov-Chou-Mughees demonstrate that these techniques intentionally preserve substantial signal and, by design, “many applications of sensor data are unlikely to be affected” (Section 7, CONCLUSION).

Regarding battery life, obfuscation, quantization, and frequency reduction use the same underlying sensors, so they plainly do not resolve the problem.

Can you cite some evidence to support your assertion that obfuscation, quantization, and frequency reduction solve the privacy and consent problem?

Or, if the solution is as simple as you say it is, you might as well just code it up and we can test it empirically.

Until then, we have an absence of evidence to support your assertion, and a presence of evidence to refute your assertion, so we have to conclude that your assertion is without merit.

**Comment by TimvanScherpenzeel** [comment]
For any new readers: earlier in this thread the new study on https://sensor-js.xyz/ is referenced. The relevant paper can be found here: https://sensor-js.xyz/webs-sixth-sense-ccs18.pdf. From what I understand the main issue is the misuse of the sensors for browser fingerprinting and various ad-tracking techniques (such as distinguishing between bots and real users).

In the study some other studies and exploitations are referenced (reading PIN's, reading keystrokes, sniffing browser history) which many have already said in the past to be highly impractical and extremely difficult to pull off in a short amount of time in the confined space that is the web browser. They are neat hacks (see: https://blog.lukaszolejnik.com/stealing-sensitive-browser-data-with-the-w3c-ambient-light-sensor-api/) but I have problems considering them to be actual threats.

The paper is also not clear on how this raw data is actually used to help towards browser fingerprinting seeing as the sensor data rapidly changes over time and is not a consistent identifier over a longer period of time. If someone could explain that to me in more detail it would be highly appreciated.

I think there is plenty we can do to make it a lot more difficult for advertising networks to misuse these sensors (for example block any 3rd party iframe embeds / cross-origin sources from using these sensors by default, even secure sources). Browser fingerprinting is still very successful and won't be affected in any major way by this change: just have a look at fingerprinting through WebRTC / WebSocket / canvas / font detection / plugin support / evercookie / user agent / WebGL renderer & extensions. 

The damage of this proposed change far outweighs the possible benefits in my opinion and should be seen as a last resort. I feel that we should push for blocking default sensor access through 3rd party services (even secure ones) but leave it open for 1st party users that can use these sensors in creative ways (AR / VR / parallax effects etc..).

**Comment by othermaciej** [comment]
@TimvanScherpenzeel 
- We (Apple) consider all of the cited threats to be real threats. We trust our  security experts

- We consider first-party fingerprinting to be a threat, not just third-party fingerprinting.

- We don't accept the argument that fingerprinting attack surface can be added freely because other fingerprinting attack surface already exists. We agree with the [TAG Security and Privacy Questionnaire's view on sensor data](https://w3ctag.github.io/security-questionnaire/#sensor-data). (In any case, there are risks more severe than fingerprinting.)

Given these considerations, it's unlikely we will enable this API by default on Apple platforms. We will however consider enabling it via user consent if there's agreement on a permissions API (which there n=mostly seems to be).

**Comment by TimvanScherpenzeel** [comment]
Thank you for your clear reply @othermaciej.

In that case I trust that you and the security experts are making the right choice in making this change. I don't have the experience in what it takes to make a browser safe nor do I have a clear view on exactly how malicious parties are exploiting the mentioned sensor API's. The [Gyrophone](https://www.usenix.org/system/files/conference/usenixsecurity14/sec14-paper-michalevsky.pdf) reference in the TAG Security and Privacy Questionnaire is really clever.

I hope we can find an elegant solution for how the requesting of permissions is done. To be completely honest, when I started looking into the API's myself about a year and a half ago I was very surprised users did not have to give explicit permission.

**Comment by tangobravo** [comment]
> According to an [article](https://medium.com/@firt/pwas-on-ios-12-2-beta-the-good-the-bad-and-the-not-sure-yet-if-good-a37b6fa6afbf#1e1f) discussing iOS 12.2 beta features:
> 
> > * Accessing sensor APIs is now disabled by default, so no accelerometer, gyroscope or magnetometer APIs are working unless the user enables a Settings option under “Safari”
> 
> Assuming this beta feature will ship in stable (edit: it will according to the [release notes](https://developer.apple.com/documentation/safari_release_notes/safari_12_1_release_notes)), 

[...]

I've seen quotes from earlier in the day from the release note page including a sentence that the option to allow sensor events would be disabled by default. Looking at the link now, there is no comment about the default state of these options. @anssiko can you confirm this was explicitly specified before? Anyone from the Safari team care to clarify?

The worry of this change is it buries the option in the settings without even offering a permissions API to the browser. I'm on the AR/VR use case side and think a permission prompt perfectly acceptable. However having to direct users to some obscure part of the settings app to enable this option definitely seems a step too far which will be damaging for the web.

My hope is Apple/Safari have decided to keep the setting enabled by default in 12.2. That will give time to finalise the permission model without breaking web content, but still give the privacy conscious the opportunity to disable globally.

**Comment by griffin2000** [comment]
The key point here, as someone who is building their technology and  business model on these web standards, is this needs to be resolved quickly.

Personally I feel that a permission pop-up with the existing API is the way to go.  But if that's not the solution you go with, that's fine.  As long as the result is that our customers, who visit our website can actually use it, as specified by W3C standard we trusted, and dedicated a lot of engineering and design time to target.

**Comment by othermaciej** [comment]
Just to be clear about the current state of the beta: the feature is currently enabled by default everywhere unless the user manually turns on the user preference. If we had per-site permission, then that would take priority over the setting.

I don't think we'd piggyback this on permission for another API. That's likely to be confusing to users. Details on what Apple can or should do are probably better reported on bugs.webkit.org than hear though. I'm offering updates only to set context on our need for the Permission API.

**Comment by othermaciej** [comment]
@griffin2000 Yeah, I did. Corrected in my original post.

**Comment by griffin2000** [comment]
Got it.  This is really not an acceptable outcome.  Disabling parts of the released, widely used, parts of web standard is not going to lead to this technology being accepted and used in widespread manner.  At this point, we would have been better off dedicating our development effort towards supporting these features in a native app.  We naively thought basing out solution on web technologies would lead to a more widely available solution. 

Added a WebKit bug here, for reference
https://bugs.webkit.org/show_bug.cgi?id=194183

**Comment by zchrykng** [comment]
> widely used

Not that I doubt you on this, but are there any statistics showing how much this is in use? I personally have never run into a site that used any of these features.

**Comment by othermaciej** [comment]
We acknowledge the value of these things and we wish we had a solution that both respects privacy and provides maximum user convenience. Our first priority has to be to fix the privacy exploit though.

**Comment by richtr** [comment]
@geoffreygaren In reply to [this comment](https://github.com/w3c/deviceorientation/issues/57#issuecomment-459818413):

> Can you cite some evidence to support your assertion that obfuscation, quantization, and frequency reduction solve the privacy and consent problem?

Are you seriously suggesting that, returning e.g. integer values for `deviceorientation` is going to allow an attacker to identify a single device/user from a database of four billion smartphone devices? 

Everything being discussed is a shade of that argument that there is, almost certainly, a level of default access available that provides enough privacy and security protections while not completely disabling current uses.

**Comment by tangobravo** [comment]
@othermaciej Thanks for the responses here.

> Just to be clear about the current state of the beta: the feature is currently disabled by default everywhere unless the user manually turns on the user preference. If we had per-site permission, then that would take priority over the setting.

Thanks for clarifying the current state. For the full 12.2 release you have (at least) 3 options:
1) Keep things as in the beta: many sites stop working, not many users would go through the hassle of turning on the option, horrible UX to talk users through the process.
2) Switch the default setting to be enabled: gives time for more discussion whilst still allowing privacy-conscious users to turn it off. [I'd throw in implementing other less-impactful mitigations in this option too, but given timeframes I think that is unlikely.]
3) Keep the default disabled but implement a permission request API: "progressive enhancement" use cases likely break, but those that require this API (360 photos etc) would be able to adapt.

**I would strongly advocate for option 2 or 3.**

Sorry for hijacking the bug for continuing this discussion. WebKit (rightly IMHO) don't consider it their bug, and the private nature of Apple's bug tracker shuts down discussion and prevents the web community from demonstrating the strength of feeling on this issue.

**Comment by richtr** [comment]
FYI, I will now recuse myself from further discussion here to allow others to make progress on selected proposals (unless pinged directly for input or clarification on previous points).

**Comment by geoffreygaren** [comment]
> Are you seriously suggesting that, returning e.g. integer values for `deviceorientation` is going to allow an attacker to identify a single device/user from a database of four billion smartphone devices?

"returning e.g. integer values for `deviceorientation`" is an incomplete specification. 

What I'm suggesting is that you have provided no evidence that you could complete that specification in a meaningful way. And I have provided evidence that you can't.

If the solution is indeed trivial, you are welcome to end this debate at any time by specifying the solution and providing evidence that it works.

> Everything being discussed is a shade of that argument that there is, with certainty, a level of default access available that provides enough privacy and security protections while not completely disabling current uses.

Indeed, you do keep asserting that you could solve the privacy and consent problem. But asserting something doesn't make it so. Evidence is how we evaluate truth, and you have provided none.

**Comment by tparisi** [comment]
Two questions: 1) is this same change going to be required of native apps going forward? If not, why? 2) has anyone considered all the non-WebXR content this will break, like HTML5 games or websites that use orientation to change layout?

**Comment by griffin2000** [comment]
> Apple/WebKit aren't going to change their minds on this. iOS 12.2 is going to ship with device orientation/motion disabled by default. We need to expedite the process of formalizing an API so that future versions of iOS can ship with it. If we move quickly, maybe even iOS 12.2 could have this implemented.

This is by far the most important aspect of this discussion.  The exact method that happens FAR is less important than coming up with a solution that can get this fixed for the 1.2.2 full release.

If that fails it will be really bad for web standards as a whole, and, speaking as a developer, make me much less likely try and support these kind of standards in the future.

**Comment by grorg** [comment]
> Two questions: 1) is this same change going to be required of native apps going forward? If not, why?

Possibly, yes. But it's a different situation - web sites are including third-party code that is violating privacy by using this API. Apps typically don't do that (as often).

And either way, the privacy situation exists on the Web, so it needs to be solved here.

> 2) has anyone considered all the non-WebXR content this will break, like HTML5 games or websites that use orientation to change layout?

Websites shouldn't use the deviceorientation API to change layout. There are better ways to do that.

Yes, some games will be affected. As will things like panoramic photo viewers (not sure if they count as XR).

**Comment by reillyeon** [comment]
With my working group chair hat on: To new folks coming to this issue, let's try to avoid piling on and making this an unfriendly place to collaborate. We have engagement from three implementers on this thread and a proposed specification change should be available soon.

**Comment by cdumez** [comment]
Another thing that came up while discussing this internally is the need for a user gesture in order to call requestPermission(), similarly to what is done for getUserMedia. Any thoughts on this?

This would limit potential abuse.

**Comment by reillyeon** [comment]
👍 Displaying a permission prompt should require a user gesture.

**Comment by KenanSulayman** [comment]
Requiring a user gesture seems reasonable, although, does explicitly navigating to a site count as user gesture?

**Comment by machenmusik** [comment]
A few thoughts

- It seems clear that treating device motion and orientation events differently than other user inputs such as camera and microphone was unfortunate in hindsight.

- It seems logical to effectively promote them to user input with respect to privacy considerations, and reuse the same affordances used for other user inputs such as camera and microphone here.

- Ideally this would be a default stance taken to protect user privacy, as I am not sure exactly  how/if this would impact generic sensors and other less established functionality at present.

**Comment by tobie** [comment]
We did a lot of work around these issues in the Generic Sensor API. It's worth having a look at the [Security and Privacy section](https://w3c.github.io/sensors/#security-and-privacy) of the spec, and some [specific conversations](https://github.com/w3c/sensors/issues?utf8=%E2%9C%93&q=is%3Aissue+label%3Aprivacy) in the tracker.

@lknik and @maryammjd might have more to add.

**Comment by samuelweiler** [comment]
I am glad to see that a rough consensus has emerged.  If implementers want feedback on specific choices (e.g. user gesture before prompt), you are welcome to ask the participants in the [W3C Workshop on Permissions and User Consent](https://www.w3.org/Privacy/permissions-ws-2018/papers.html) held in September.  You can reach the lot of us at: public-2018-permissions-ws@w3.org

**Comment by erikmchut** [comment]
Just some food for thought as progress is made here. Native apps and the web alike are moving towards more natural UX, where moving and pointing your phone is how you interact with the content. ARKit (native), ARCore (native), 8th Wall (web), and AR.js (web) are prominent examples of this.

Put another way, device orientation and motion are starting to replace user touches as input.

On native, Apple has made this very easy with ARKit.
On web, iOS Safari is about to make this harder or default-disabled.

Please consider the implications of a double-standard that indirectly benefits a closed app ecosystem. 

Also, as you make changes here, do please consider the following:

- Unlike other mobile browsers, Safari/WebKit doesn’t currently cache permissions per-domain, so any permission request occurs *every single time* you view the page. This contributes to permissions fatigue and users generally making less informed decisions about privacy.
- For pages whose entire input is device motion / orientation, like many AR/VR pages, there is currently no user gesture (since canvas touches aren’t considered user gestures). Thus in the pathologically worst end-state, Safari might require a prompt for each of the following *every* pageload:
    - 1) Bespoke dialog stating ‘This page would like to use camera and device motion, click ‘here’ to request permission.
    - 2) Permission dialog, this page is requesting camera access (yes/no)
    - 3) Permission dialog, this page is requesting device motion (yes/no)
    - 4) Permission dialog, this page is requesting device orientation (yes/no)

While I expect that we end up with a much better state than the scenario above – If I could suggest anything, it would be to enable permissions caching at the domain level in Safari, perhaps combined with a check-box that says ‘don’t ask again for this site’. This would prevent permissions fatigue and give parity with native apps which do not need to request permission on every app load. Much of this may be addressed in future permissions APIs, but it would be helpful if Safari/WebKit teams would be open to mitigate these concerns before making changes to device motion/orientation permissions.

**Comment by lknik** [comment]
Hi,

I wrote a bit on sensors a while ago (ambient light [1](https://blog.lukaszolejnik.com/stealing-sensitive-browser-data-with-the-w3c-ambient-light-sensor-api/), [2](https://blog.lukaszolejnik.com/additional-security-and-privacy-risks-of-light-sensors/)). 

Not sure if helpful in your conversation, but we also analysed the possible impact of changes [3](https://blog.lukaszolejnik.com/battery-status-not-included-assessing-privacy-in-w3c-web-standards/) ([PDF](http://lukaszolejnik.com/AssessingPrivacyWebStandardsIWPE17.pdf)).

We all know well the debate on privacy vs usability/precision, and also the point of view of whether something is or is not a threat. In the end it boils down to individual beliefs. I've seen and shown some of the risks. While I'm not saying such attacks are always 100% reliable, I also know that quantization and frequency capping often don't work (see e.g. (1)).

And in that case I'm happy to hear the position of Apple as presented by @othermaciej 
How to handle it with the "defaults" is another thing. I'm all for opt-in, the current model of the web is opt-out (with the exception of permission-gated things).

**Comment by richtr** [comment]
Replying to [this comment](https://github.com/w3c/deviceorientation/issues/57#issuecomment-459979734) from @geoffreygaren:

>> Everything being discussed is a shade of that argument that there is, with certainty, a level of default access available that provides enough privacy and security protections while not completely disabling current uses.
> 
> Can you cite some evidence to support your assertion that obfuscation, quantization, and frequency reduction solve the privacy and consent problem?

Empirical evidence for the case for obfuscation, quantization, and frequency is, I agree, important to proceed. If those approaches are capable of eliminating all privacy and security concerns then the consent problem is also taken care of implicitly. We are in the current predicament due to the lack of clarity within existing studies on the exact protections that would eliminate personally identifiable data concerns. Existing studies do not go far enough in evaluating the full effects that obfuscation, quantization, and frequency can have on eliminating personally-identifiable information from sensor data and, thus, eliminating fingerprinting. However, it remains equally disappointing to see a progressive, pro-active approach based on further investigation being dismissed in favor of a quick short-term solution.

In [Bojinov-Boneh-Michalevsky-Nakibly](https://crypto.stanford.edu/gyrophone/sensor_id.pdf), Section 6.1 states the following:

> " In contrast to audio-based fingerprinting, there is no
good way to feed a signal into the accelerometer, namely
exert a known acceleration force; instead we take an
approach of performing background measurements and
waiting until there is enough data to estimate the accelerometer
calibration parameters. We perform a measurement
every time the phone is at a resting position, or
more precisely, the phone is at a constant velocity (no acceleration).
Note that in most reasonable cases it is very
unlikely that a phone will not be at rest for an extended
period of time. Detecting the phone is at rest is relatively
straightforward: the measured acceleration vector should
be static and its magnitude should roughly be equal to `g` 
(gravity). "

Figure 6 in that same paper provides a scatter plot of all z-axis accelerometer data that is collected from 17 iOS devices when they are placed at rest on a table (i.e. the device has no velocity in any axis thus fulfilling the required calibration environment requirements as explained in the quote above). 

All the iOS devices in that study present unique `g` values that fall within in the range `[0.993, 1.004]`. That study showed that `devicemotion` data is potentially fingerprintable based on those discrepancies between different devices (although many devices reported values do overlap with each other within that same range).

In a [previous comment above](https://github.com/w3c/deviceorientation/issues/57#issuecomment-459302243) I suggested _obfuscation_ as follows should be added to emitted `devicemotion` data:

> * Add a random constant in the range ... [-0.005, 0.005] to all `devicemotion` data on each new invocation of the API.

By applying these obfuscation values to the data collected in that study, that data from 17 devices becomes indistinguishable from any subsequent invocation of the `devicemotion` API on the same device OR any invocation of the `devicemotion` API on any other device. Where the term "fingerprint" is defined as a means to identify and then re-identify a user on a subsequent visit due to consistencies in readings, this API would become useless for the purposes of ongoing identification of devices or users between visits.

Finally, in the same [previous comment above](https://github.com/w3c/deviceorientation/issues/57#issuecomment-459302243), I continued that discussion around obfuscation with the following:

> Optionally adjust that value during idle periods of inactivity for dynamic obfuscation (when e.g. the device is placed on a table).

Implementing this additional protection goes even further to prevent any fingerprinting from occurring within the same continuous, long-running stream of `devicemotion` data collected on a device. It does this by adjusting the obfuscation constant marginally (by a much smaller value) whenever the device is detected to be at rest (e.g. placed on a table or held still with no exerted velocity in any axis).

Finally, the presence or absence of any obfuscation constant in this data does not affect general calibration of `devicemotion` sensor data. Calibration of device sensors is performed under the same conditions as described above (when no velocity is present in any axis) - at which point we can estimate the calibration offset from earth's precise gravitational constant of 0.980665 (9.81 m/s^2) on each invocation.

In [this comment](https://github.com/w3c/deviceorientation/issues/57#issuecomment-459818413), @geoffreygaren wrote:

> Or, if the solution is as simple as you say it is, you might as well just code it up and we can test it empirically.

In WebKit, the most trivial code change would be to add a per-instance-bound obfuscation constant (`CURRENT_OBFUSCATION_CONSTANT`) in the range `[-0.005, 0.005]` to the collected gravitational values on [these lines](https://github.com/WebKit/webkit/blob/c3ad6ddc321cc9a1cc99016cd1ae1dff25049a10/Source/WebCore/platform/ios/WebCoreMotionManager.mm#L238-L240):

```diff
- totalAccel.x = userAccel.x + gravity.x;
+ totalAccel.x = userAccel.x + (gravity.x * (1 + CURRENT_OBFUSCATION_CONSTANT));
- totalAccel.y = userAccel.y + gravity.y;
+ totalAccel.y = userAccel.y + (gravity.y * (1 + CURRENT_OBFUSCATION_CONSTANT));
- totalAccel.z = userAccel.z + gravity.z;
+ totalAccel.z = userAccel.z + (gravity.z * (1 + CURRENT_OBFUSCATION_CONSTANT));
```

A basic patch of this type would provide enough initial protection to eliminate the majority of fingerprinting capabilities of this API being discussed.

---

By applying just 'obfuscation' initially would indicate effective removal of the ability to fingerprint devices/users via `devicemotion` if those fingerprinting attack vectors rely on any form of "at rest" calibration process. Other attack vectors (such as measuring the sensitivity of sensor data readings) could be mitigated easily with quantization protections.

The entire concept of [spatial computing](https://medium.com/@victoragulhon/what-is-spatial-computing-777fae84a499) on the web is at stake (i.e. motion as a basic input driver to rival mouse, keyboard and touch input drivers). Basic input drivers should not be subject to permission-based access controls (by all means, all other sensor APIs belong behind permissions but absolutely not motion motion and orientation sensor access).

Regarding battery life, a reduction in the frequency of event firing would materially reduce the amount of event collection and propagation required within a web browser's code. Execution of those code paths would be reduced by a factor of 50%+ (e.g. when moving from firing of events from 60 FPS to 30 FPS). Anything >= 15 FPS is still entirely acceptable for the use cases documented within this specification (i.e. non-VR/AR use cases). For use cases that require high-resolution data (at greater accuracy and delivered at higher frequencies) such as the use of motion and orientation sensor data within VR/AR scenes there are other, more suitable, APIs that should be implemented, gated with permissions and used by web developers (Generic Sensors API, WebVR API, WebXR API). Those APIs, obtained via permissions, should remain unencumbered with any forms of obfuscation, quantization, and frequency capping or limiting.

All of this is to say: we do still need more _independent_ feedback on the effectiveness of these mitigations before pushing existing and future spatial computing use cases off a cliff. The most reasonable approach at the current time would be for Apple to implement a simple form of obfuscation as explained above in WebKit code and to also enable the "Motion and Orientation Sensors" setting by default with those protections implemented in an upcoming release. This would provide initial protections and mechanisms to eliminate the most egregious fingerprinting attack vectors, as identified in existing sensor privacy studies, while still allowing on-going research to continue in to additional protections that could be added on an ongoing basis in the future. This approach allows Apple and other browser vendors to continue to research, tweak and adapt sensor privacy protections in future releases while still actively preventing fingerprinting in the immediate short-term. In any other approach we are simply stating that spatial computing use cases on the web are immaterial to this discussion, spatial computing as a concept and basic input has no place or future on the web and spatial computing use cases will be eliminated as collateral damage when building more walls around access to this data instead of focussing on eliminating personally-identifiable features from that data instead. The stakes of this discussion and the course of action being taken here to add permissions or otherwise restrict access to data _warrants extreme caution_. Nobody will stop browser vendors from their current course of action but I simply wish that the impact and loss to open motion access to the web was clearly understood in the context of the elimination of spatial computing uses (to the benefit of native applications).

**Comment by lknik** [comment]
> > * Add a random constant in the range ... [-0.005, 0.005] to all `devicemotion` data on each new invocation of the API.

Indeed, I'm not sure how long would that hold, considering the data range you stipulate. How many observations need to be made to discard the random noise?

**Comment by tangobravo** [comment]
@richtr - Although I agree that fingerprinting concerns are likely to be solvable with some of the mitigations you mention that is not the only potential privacy concern.

To take one example; even heavily quantized, obfuscated, and reduced-frequency sensor data would give some information about whether the user is standing up or lying down. Whether that is a significant enough privacy concern to require explicit user consent is certainly debatable (most vendors appear not to think so at the moment), but it does appear Apple / WebKit have made their minds up here.

I believe spatial computing use cases on the web would survive a permissions prompt. It would be unfortunate however to have a widely-released iOS version with the API disabled by default **and** without a permissions API implemented.

So I still hope for the final iOS 12.2 release Apple will either have an opt-in permissions API implemented or will decide to flip the default setting until that permissions API is ready. To me that strikes a good balance between addressing privacy concerns but without being quite as hostile to web developers who depend on these features.

**Comment by richtr** [comment]
@lknik wrote:

>> > * Add a random constant in the range ... [-0.005, 0.005] to all `devicemotion` data on each new invocation of the API.
>
> Indeed, I'm not sure how long would that hold, considering the data range you stipulate. How many observations need to be made to discard the random noise?

This is the purpose of quantization to sit above random noise thresholds.

**Comment by lknik** [comment]
@richtr yep, it's catchy but choose the ransom noise level needed. And thanks @tangobravo for resurfacing that privacy is not merely about fingerprinting. Actually many of the risks I considered do not relate to fingerprinting at all. Sensors is such a rich data source...

**Comment by neave** [comment]
@cdumez @reillyeon @othermaciej @anssiko What is the process now for agreeing upon an API and adding it to the official spec? I think most of us would like to see this discussion moved on and to implementing a solution for developers.

**Comment by anssiko** [comment]
Procedural steps need to be taken before a contribution can be added to a W3C WG spec from outside the group's current participants. This process is on the fast track and involved parties are informed.

The chairs will update this issue when we have new information to share. Meanwhile, technical discussion can continue on this issue. I can assure you everyone is highly motivated to resolve this issue. Thank you for your understanding.

**Comment by machenmusik** [comment]
In reply to:

```
The entire concept of spatial computing on the web is at stake (i.e. motion as a basic input driver to rival mouse, keyboard and touch input drivers). Basic input drivers should not be subject to permission-based access controls (by all means, all other sensor APIs belong behind permissions but absolutely not coarse-grained motion and orientation sensor access).
```

My view is that the reason why we are looking for a different solution than default-enabled exposure is precisely because, for the browsers in question like Mobile Safari, the spatial computing aspect is not likely to be a primary function, and therefore this data is not likely to be a primary input source.

As an example, if Avaer were to decide that ExoKit is intended to be an immersive browser with spatial computing as a primary function, then ExoKit could decide to default to enabled for https sites or localhost (or perhaps to any site displayed in immersive mode, if that were too restrictive).

My intuition is that obfuscation of the data is at this moment an uncertain solution, and that a more intentional permission model will be needed, with browsers ideally remembering site permissions where appropriate to reduce user friction.

**Comment by nuthinking** [comment]
With all due respect, it looks like a lazy solution to me. Please let’s stop showing people popups and let’s find instead ways to prevent this information to cause any harm. Didn’t we learn anything from the cookies banners? Advertisers will track everything. If you don’t want websites to know your device orientation, stay away from the Web. MUCH WORSE THINGS HAPPEN OVER THERE! ✌️

**Comment by machenmusik** [comment]
@not thinking If you scroll upwards, you will see that there are several discussing work to mitigate potential harm, but also significant discussion that current techniques may not be sufficient.  That discussion has been going on for some time, and appears likely to continue for some time.

In the meantime, I think it makes sense to do something additional to mitigate, and the obvious  choices are to either disable by default with no interactive reprieve (that entirely kills the use cases, which is why folks are looking for other solutions here) or to treat like other sensitive inputs with interactive opt-in (camera, microphone).  Those inputs cannot generally be degraded or rate limited without significantly impacting their primary use cases, and I suspect the same may prove true here.

**Comment by nuthinking** [comment]
@machenmusic I am not a security expert and I appreciate being conservative when it comes to security. I would just hope the discussion was focused on the security aspect. Why having throttling, or a way to temporary disable the feature, for instance, while typing in a textfield marked as password, wouldn’t work? I don’t buy the privacy issues. That data Is not as exploitable as mic and camera. Should we ask permissions to get the mouse positions between clicks then? Can we clearly state the problem we want to solve? Not the solution.

**Comment by machenmusik** [comment]
@nuthinking this has been discussed above - this is one of the longer threads that I can remember, actually - and I think the browser implementers (i.e. the ones with exposure risks and real skin in the game) have gotten at least as far as https://github.com/w3c/deviceorientation/issues/57#issuecomment-459537029.

**Comment by nuthinking** [comment]
Privacy hole... c’mon. You can probaly tell if someone is in bed by how they scroll the device. So what? Who defines the privacy line? Safari is disabling the feature and user will have to enable it in the settings (maybe gamers will do it).

**Comment by reillyeon** [comment]
This discussion is becoming unproductive. I am locking the conversation until a PR for the proposed API changes has been posted.

**Comment by anssiko** [comment]
A PR https://github.com/w3c/deviceorientation/pull/68 to address this issue has been submitted for review.

Thanks @cdumez!

**Comment by machenmusik** [comment]
Thank you!

**Comment by maryammjd** [comment]
This is good news for the security and privacy research community. Does the DAS WG plan to go for the same approach for other sensors too?

**Comment by reillyeon** [comment]
@maryammjd, see w3c/sensors#388.

**Comment by nbutko** [comment]
Thanks for updating this issue, it's great to see the progress here. I was wondering it it's possible to get some minor points clarified related to the proposed specification PR #68 

Based on the spec, the user interface would look something like:

1. A bespoke dialog to approve motion and orientation permissions to view in AR with an 'Ok, got it' button (explicit action), clicking this button fires two promises that result in two prompts:
2a. Prompt for device motion for origin
2b. Prompt for device orientation for origin
3. Then start.

Is there a mechanism to check if permission is granted prior to (1)? It may be confusing to see 'Permission is required, please allow' followed by no prompts (immediately return 'granted').

Are implementations allowed to merge (2a) and (2b) into a single prompt (2)? If so, adding something to this effect ("Implementations may choose to resolve simultaneous permissions promises with a single user prompt") could help clarify this.

Any clarification on these points would be greatly appreciated.

**Comment by cdumez** [comment]
@nbutko In WebKit at least, I have currently implemented it as a single prompt, permissions are merged on native side. Web pages should still request permission for both types of events if they need both (the second promise will be resolved without prompting in WebKit). I am personally not sure the spec needs clarification in this respect, it is up to browsers to decide how they want to do prompting.

**Comment by nbutko** [comment]
Thanks @cdumez that's great to hear. 

Any guidance on checking of the permission state is 'granted' prior to a custom prompt for permission?

**Comment by cdumez** [comment]
@nbutko This is a good point. My proposal would be:
1. The page calls requestPermission() without user gesture.
2.a. If it is rejected, then you do not have permission and you need to request again with a user gesture
2.b. If it gets resolved with 'granted', then you already have permission
3.b. If it gets resolved with 'denied', then the user has already denied access in the past.

If this is acceptable, I would need to clarify the spec text and update the WebKit implementation accordingly.

**Comment by nbutko** [comment]
That would work well. Thanks!

**Comment by machenmusik** [comment]
Excellent clarifications, thanks again.  @cdumaz @reillyeon @anssiko should any further discussion remain here, or move to #68?

**Comment by cdumez** [comment]
> Excellent clarifications, thanks again. @cdumez @reillyeon @anssiko should any further discussion remain here, or move to #68?

I think on PR #68 would be better now assuming it is about the proposed change itself.

**Comment by nbutko** [comment]
@cdumez Will you plan to update this thread again when the webkit implementation of #68 is live, enabled and available to try in iOS beta (if it isn't already)?

**Comment by cdumez** [comment]
@nbutko Yes, I can update this thread once the implementation is live in an iOS beta build. It is definitely not yet in an iOS beta given there was no PR or feedback until today :)

**Comment by nbutko** [comment]
@cdumez I noticed that a release version of 12.2 was pushed out today. Does this have your changes to make webkit conform to the recently added request permissions spec?

**Comment by cdumez** [comment]
> @cdumez I noticed that a release version of 12.2 was pushed out today. Does this have your changes to make webkit conform to the recently added request permissions spec?

No, it doesn't. I cannot comment on *when* it will ship but I have already agreed to post a message on this issue when the permission API becomes available in a public beta.

**Comment by neave** [comment]
> No, it doesn't. I cannot comment on _when_ it will ship but I have already agreed to post a message on this issue when the permission API becomes available in a public beta.

This is disappointing. I hope the API will make it into an update soon.

For anyone looking for a workaround, the best thing to do is to detect the disabled state via a timeout and clear it in the event handler. Something like this:

```
const id = setTimeout(() => alert('Enable device orientation in Settings > Safari > Motion & Orientation Access.'), 500);

window.addEventListener('deviceorientation', event => {
	clearTimeout(id);
	// ...
});
```

**Comment by keithito** [comment]
To add to @neave's workaround, it appears the deviceorientation API still works in Chrome, so you could prompt the user to open the page in Chrome if you detect that deviceorientation is broken. Something like:

```
<div id="nodeviceorientation" style="display:none">
  This page no longer works in your browser.<br>
  <a href="googlechromes://YOUR_URL">Open in Chrome</a> or 
  <a href="itms-apps://itunes.apple.com/us/app/chrome/id535886823">Install Chrome</a>
</div>

<script>
const id = setTimeout(() => {
  document.getElementById('nodeviceorientation').style.display = 'block'
}, 500);
window.addEventListener('deviceorientation', (e) => clearTimeout(id));
</script>
```

Not a great solution, but possibly the easiest option for the user given the current state of things.

**Comment by EvanBacon** [comment]
@cdumez it'd be nice if toggling the switch at `Settings > Safari > Motion & Orientation Access.` would reset the Safari app, this would be analogous to toggling native app permissions.
Otherwise the developer would need to observe the app state and attempt to resolve motion access again.

**Comment by micsun-al** [comment]
Fun edge case...
**iPad Air (MD785LL/A) + iOS 11.4:**
  Does not emit either devicemotion or deviceorientation events if the tablet is laying flat on it's back (on a table or whatever).

**iPhone X + iOS 12.2 + "Motion and Orientation Access" on:**
  Laying on a table does emit both events.

I guuuesss we can add a some window.navigator.userAgent checks, but getting in to the "pretty janky checks" arena :\

**Comment by reillyeon** [comment]
Regarding the last couple comments, please report implementation-specific issues to the appropriate issue tracker.

**Comment by nbutko** [comment]
@cdumez Does iOS 12.3 add an implementation for the request permissions spec?

**Comment by reillyeon** [comment]
For WebKit and iOS questions please use [bugs.webkit.org](https://bugs.webkit.org) and [bugreport.apple.com](https://bugreport.apple.com) respectively. This is not the right place to ask about implementation status in particular engines.

**Comment by nbutko** [comment]
@reillyeon is there a link to an existing bug that we can track progress and get updates on?

**Comment by reillyeon** [comment]
I searched bugs.webkit.org but didn't find one. @cdumez might be able to provide a link.

**Comment by cdumez** [comment]
Bugs.webkit.org is not a way to track when a particular feature ships. Apple also does not comment on when a particular will ship.

@nbutko Please coordinate internally. I already asked Erik from 8th Wall to file a bug on bugreport.apple.com and he did (49748962). When the feature ships, this bug should get closed and you will be notified.

You could also simply test if DeviceOrientationEvent.requestPermission is defined in any new build that comes out (it is *not* in 12.3), or even ping me on Twitter (@chris_dumez) where I am already in contact with 8th Wall. 

Just please let's stop pinging on this GitHub issue.

**Comment by cdumez** [comment]
For people interested in trying out this new API on iOS, I suggest installing the iOS 13 developer beta 1. If you find issues, please file them at https://bugs.webkit.org, not here. I am also happy to answer questions on Twitter (@chris_dumez) or at WWDC WebKit Labs this week for those attending.

**Comment by ad17107** [comment]
hello paul good night

**Comment by am17107** [comment]
i can help

**Comment by am17107** [comment]
ok

**Comment by cdumez** [comment]
This is not the place for WebKit support. Please file bugs on bugs.webkit.org or ping me on Twitter.
WKWebView currently has access the device motion and orientation without permission so we did not expose the permission API there. This may change in the future though.

**Comment by elmariocarlos** [comment]
Hi @chris_dumez I filed the bug here, let me know if there is any further specification that should be looked at: https://bugs.webkit.org/show_bug.cgi?id=203287

**Comment by mvl22** [comment]
Useful links for those wanting to implement this in their client code, to save you searching:

https://dev.to/li/how-to-requestpermission-for-devicemotion-and-deviceorientation-events-in-ios-13-46g2

https://medium.com/flawless-app-stories/how-to-request-device-motion-and-orientation-permission-in-ios-13-74fc9d6cd140

**Comment by gkoberger** [comment]
Just a quick note for anyone who keeps running into `deny`s without seeing a prompt; you need to be in `https` or it won't work! I wasn't able to get this working locally for this reason, but it works otherwise.

**Comment by solarsu** [comment]
Hi @cdumez I have the same problem
1.when I use UIWebview in ios13 iphone and and the address is https://www.audero.it/demo/device-orientation-api-demo.html  it can work
2.but when I use WKWevView in the same iphone and same address it can't work 
3.when I use UIWebview in ios12.3 iphone and and the address is https://www.audero.it/demo/device-orientation-api-demo.html  it can work
how can I get the the device motion and orientation without permission in ios 13 by WKWebView 
ths

**Comment by yoyo837** [comment]
iOS 12.2+ (Privacy switch turned on): 

| - | http | https |
| -------- | ----------- | ---- |
| Safari | No | Yes | 
| WebView | No | Yes | 

iOS 12.2+ (Privacy switch turned off): 

| - | http | https |
| -------- | ----------- | ---- |
| Safari | No | No | 
| WebView | No | Yes |

**Comment by solarsu** [comment]
@yoyo837 thank But I want to know how  WKWebView turned on in ios13

**Comment by cdumez** [comment]
> Hi @cdumez I have the same problem
> 1.when I use UIWebview in ios13 iphone and and the address is https://www.audero.it/demo/device-orientation-api-demo.html it can work
> 2.but when I use WKWevView in the same iphone and same address it can't work
> 3.when I use UIWebview in ios12.3 iphone and and the address is https://www.audero.it/demo/device-orientation-api-demo.html it can work
> how can I get the the device motion and orientation without permission in ios 13 by WKWebView
> ths

Bugs.webkit.org is the right place for WebKit support.

Also, I have just tried https://www.audero.it/demo/device-orientation-api-demo.html in a WKWebView and it worked fine. To test, I used the WebView app in the App Store.

**Comment by reillyeon** [comment]
I'm locking this issue. Please use the WebKit issue tracker for questions about WebKit's implementation.
